### ISSUE TITLE
1a. Extract text/ module (normalization, compilation, splitting)

### ISSUE THREAD

**Issue by jakebromberg** [WXYC/wxyc-etl#1]
## Overview

Extract text normalization, compilation detection, artist splitting, and batch normalization from `discogs-xml-converter/src/filter.rs` and `discogs-cache/lib/artist_splitting.py` into `wxyc-etl/src/text/`.

## Key Deliverables

- `normalize_artist_name()` / `strip_diacritics()` — NFKD + combining removal + lowercase + trim
- `is_compilation_artist()` + `COMPILATION_KEYWORDS`
- `split_artist_name()` / `split_artist_name_contextual()`
- `ArtistFilter` / `TitleFilter` structs
- `batch_normalize()` for PyO3 batch operations

## Dependencies

None — can start immediately.

## Plan

See `plans/phase-1-rust-etl-crate/1a-text-module.md` for the full implementation plan including TDD strategy.

---

## Implementation Plan

# 1a. Extract `text/` module into wxyc-etl

## Overview

Extract text normalization and filtering primitives from `discogs-xml-converter/src/filter.rs` into a shared `wxyc-etl` crate. This module becomes the single source of truth for artist name normalization, compilation detection, and artist splitting across all WXYC repos (Rust native + Python via PyO3 in task 1g). The existing code already implements NFKD decomposition + combining-character removal + lowercase + trim, matching the Python `filter_csv.py:normalize_artist()` behavior. The extraction replaces hardcoded Unicode combining ranges with the `unicode-categories` crate for completeness.

## Prerequisites

None. This is a foundational task with no dependencies. The `wxyc-etl` crate must be created as a new Rust workspace member.

## Pre-Implementation Dependency Check

This task has no prerequisites, but its outputs are consumed by many downstream tasks. Before starting, confirm no earlier work has created a conflicting `wxyc-etl` crate structure. If a `wxyc-etl/Cargo.toml` already exists from a parallel task (1b, 1c, 1d), coordinate to avoid overwriting each other's module declarations in `lib.rs`.

| Prerequisite | What to verify | How to verify |
|---|---|---|
| No conflicting crate scaffolding | `wxyc-etl/` directory does not yet exist, or if it does, `src/lib.rs` does not declare a `text` module | `ls wxyc-etl/src/lib.rs` and check for `mod text;` |

### Crate scaffolding

- `wxyc-etl/Cargo.toml` -- workspace crate with `text` feature flag
- `wxyc-etl/src/lib.rs` -- top-level module declarations

### Text module

- `wxyc-etl/src/text/mod.rs` -- module declaration
- `wxyc-etl/src/text/normalize.rs` -- core normalization functions:
  - `pub fn normalize_artist_name(name: &str) -> String` -- NFKD + combining removal + lowercase + trim (port of `filter.rs:normalize_artist()`)
  - `pub fn strip_diacritics(s: &str) -> String` -- NFKD + combining removal only (no lowercase/trim), for title normalization and other contexts
  - `fn is_combining(c: char) -> bool` -- uses `unicode-categories` crate (`UnicodeCategories::is_mark()`) instead of hardcoded ranges
  - `pub fn normalize_title(title: &str) -> String` -- delegates to `normalize_artist_name()` (same behavior, different semantic name)
- `wxyc-etl/src/text/compilation.rs` -- compilation/various-artists detection:
  - `pub const COMPILATION_KEYWORDS: &[&str]` -- `["various", "soundtrack", "compilation", "v/a", "v.a."]` (from discogs-cache `lib/matching.py`)
  - `pub fn is_compilation_artist(name: &str) -> bool` -- checks lowercased name against keyword set
- `wxyc-etl/src/text/split.rs` -- multi-artist name decomposition:
  - `pub fn split_artist_name(name: &str) -> Option<Vec<String>>` -- split on `/`, `&`, `and`, `with`, `feat.`, etc. (port of discogs-cache `lib/artist_splitting.py:split_artist_name()`)
  - `pub fn split_artist_name_contextual(name: &str, known_artists: &HashSet<String>) -> Option<Vec<String>>` -- context-aware splitting that only splits when components are known artists (port of `split_artist_name_contextual()`)
- `wxyc-etl/src/text/filter.rs` -- HashSet-based artist/title filtering:
  - `pub struct ArtistFilter` -- normalized HashSet with optional alias support (port of `filter.rs:ArtistFilter`)
  - `pub struct TitleFilter` -- normalized HashSet with bracket-suffix stripping (port of `filter.rs:TitleFilter`)
- `wxyc-etl/src/text/batch.rs` -- batch operations for PyO3 efficiency:
  - `pub fn batch_normalize(names: &[String]) -> Vec<String>` -- normalize a batch of artist names in one call
  - `pub fn batch_filter(names: &[String], filter: &ArtistFilter) -> Vec<bool>` -- normalize + check membership for a batch

## Dependencies

```toml
[dependencies]
unicode-normalization = "0.1"
unicode-categories = "0.1"    # replaces hardcoded combining ranges
csv = "1.3"                   # for ArtistFilter::load_aliases
anyhow = "1"

[dev-dependencies]
tempfile = "3"
```

### Cycle 1: `normalize_artist_name()` (Red-Green-Refactor)

**Red**: Write a `#[cfg(test)] mod tests` block in `normalize.rs` with the 11 test cases ported from `filter.rs::tests` (the original Python parity tests):

```rust
#[test]
fn test_normalize_lowercase() {
    assert_eq!(normalize_artist_name("Radiohead"), "radiohead");
}

#[test]
fn test_normalize_bjork() {
    assert_eq!(normalize_artist_name("Björk"), "bjork");
}

#[test]
fn test_normalize_sigur_ros() {
    assert_eq!(normalize_artist_name("Sigur Rós"), "sigur ros");
}
// ... all 11 from filter.rs
```

Run `cargo test` -- all fail because the function does not exist yet.

**Green**: Implement `normalize_artist_name()` using `unicode-normalization` for NFKD and `unicode-categories` (`UnicodeCategories::is_mark()`) for combining character detection. The implementation is a single-pass collect: NFKD decompose, skip marks, lowercase, collect, trim.

**Refactor**: Extract `is_combining()` as a helper that delegates to `unicode-categories`. Verify the `unicode-categories` crate covers all combining mark ranges (Mn, Mc, Me) that the hardcoded ranges in the original `filter.rs` cover, plus the ones they missed.

### Cycle 2: `strip_diacritics()` standalone

**Red**: Test that `strip_diacritics("Björk")` returns `"Bjork"` (preserves case) and `strip_diacritics("  Café  ")` returns `"  Cafe  "` (preserves whitespace).

**Green**: Implement as NFKD + combining removal without lowercase or trim.

**Refactor**: Factor the NFKD + combining-removal core into an internal helper shared by both `normalize_artist_name()` and `strip_diacritics()`.

### Cycle 3: `is_compilation_artist()`

**Red**: Test `is_compilation_artist("Various Artists")` returns true, `is_compilation_artist("v/a")` returns true, `is_compilation_artist("Autechre")` returns false, `is_compilation_artist("VARIOUS")` returns true (case-insensitive).

**Green**: Implement with a lowercased contains-check against `COMPILATION_KEYWORDS`.

**Refactor**: Ensure the keyword set is checked with `starts_with` or exact match semantics matching the Python implementation (the Python version checks `name.lower().strip() in COMPILATION_KEYWORDS`).

### Cycle 4: `TitleFilter`

**Red**: Port all 5 `TitleFilter` tests from `filter.rs` -- `from_file`, exact match, bracket suffix stripping, empty filter.

**Green**: Implement `TitleFilter` with normalized HashSet and bracket-suffix stripping logic.

**Refactor**: No changes expected; the logic is straightforward.

### Cycle 5: `ArtistFilter` with aliases

**Red**: Port all `ArtistFilter` tests from `filter.rs` -- `from_file`, `matches_any`, `load_aliases`, `matches_any_with_ids`, canonical name match.

**Green**: Implement `ArtistFilter` with normalized HashSet, alias CSV loading, and dual-path matching (canonical name vs. alias lookup by artist_id).

**Refactor**: Consider whether `ArtistFilter` should be `Send + Sync` (it already is in the original since `HashSet` and `HashMap` are `Send + Sync`). Ensure the parallel pipeline can share it across rayon workers.

### Cycle 6: `split_artist_name()`

**Red**: Write tests for the splitting function ported from `artist_splitting.py`:
- `split_artist_name("Stereolab / Nurse With Wound")` returns `Some(vec!["Stereolab", "Nurse With Wound"])`
- `split_artist_name("Duke Ellington & John Coltrane")` returns `Some(vec!["Duke Ellington", "John Coltrane"])`
- `split_artist_name("Autechre")` returns `None` (no separator found)

**Green**: Implement separator detection and splitting with the same separator precedence as the Python implementation.

**Refactor**: Ensure edge cases are handled (consecutive separators, separators at start/end, embedded separators in names like "Simon & Garfunkel" which should not split if context says so).

### Cycle 7: `split_artist_name_contextual()`

**Red**: Test that `split_artist_name_contextual("Duke Ellington & John Coltrane", known)` where `known` contains both normalized names returns the components, but `split_artist_name_contextual("Simon & Garfunkel", known)` where `known` does not contain "Simon" or "Garfunkel" individually returns `None`.

**Green**: Implement context-aware splitting that validates components against the known artist set.

**Refactor**: Consider whether the known set should use normalized names for comparison.

### Cycle 8: `batch_normalize()` and `batch_filter()`

**Red**: Test that `batch_normalize(vec!["Björk", "Radiohead"])` returns `vec!["bjork", "radiohead"]`. Test that `batch_filter()` returns correct boolean vector.

**Green**: Implement as simple iteration over the single-item functions.

**Refactor**: Consider whether `rayon::par_iter()` is warranted for the batch functions (likely yes for 11.5M-name batches from musicbrainz-cache, but benchmark first).

### Cross-validation integration test

Write an integration test that compares the Rust output against the Python `discogs-cache/scripts/filter_csv.py:normalize_artist()` for a set of edge-case inputs. This can be a `tests/python_parity.rs` file that hardcodes expected outputs from running the Python function, ensuring the two implementations remain in sync.

## Acceptance Criteria

- `cargo test` passes for all text module tests
- All 11 original normalization parity tests from `filter.rs` pass unchanged
- `is_combining()` uses `unicode-categories` crate instead of hardcoded ranges, and covers all Unicode General Category M subcategories (Mn, Mc, Me)
- `ArtistFilter` and `TitleFilter` have identical behavior to their `filter.rs` counterparts
- `is_compilation_artist()` matches the behavior of `discogs-cache/lib/matching.py:is_compilation_artist()`
- `split_artist_name()` and `split_artist_name_contextual()` match the behavior of `discogs-cache/lib/artist_splitting.py`
- `batch_normalize()` and `batch_filter()` provide array-in/array-out APIs suitable for PyO3 exposure
- The module is usable as `wxyc_etl::text::{normalize_artist_name, ArtistFilter, TitleFilter, is_compilation_artist, split_artist_name, batch_normalize}`
- No breaking changes to `discogs-xml-converter` (it will adopt the shared crate in Phase 3)

**Comment by jakebromberg** [comment]
Implementation PR: #13
